Improve layout of datetime columns when there's not enough space

eternoendless opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This change adds:<br>- Replace dashes with non-breaking hyphens (U+2011) in datetime default formats, so that the date is not broken up when there's not enough space.<br>- Use the appropriate date format in the logs page, according to the user's language (or configuration).<br>- Wrap the content of DateTimeColumns in `<time>` tags.
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | Yes - `PrestaShop\PrestaShop\Core\Grid\Definition\Factory\LogGridDefinitionFactory` now takes a second constructor parameter: `string $contextDateFormat`.
| Deprecations?     | no
| Fixed ticket?     | N/A - but reported by @Hlavtox 
| How to test?      | Go to a grid that has dates, then make the screen narrower. You might need to install a different language than French to get dashes, or change the date format yourself in settings.
| Possible impacts? | Might cause issues if for some reason dates are displayed using the modified code on older terminals that don't support UTF-8

Before:
<img width="355" alt="Broken date" src="https://user-images.githubusercontent.com/1009343/125348558-c81ac680-e332-11eb-91a6-754097dda934.png">

After:
<img width="262" alt="Corrected format" src="https://user-images.githubusercontent.com/1009343/125820842-94229de6-698d-4cb9-8bbd-f83456bc1c6a.png">

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/25327)
<!-- Reviewable:end -->

Commit:
Use non-breaking hyphens (U+2011) instead of dashes in date formats

Hlavtox commented:
@eternoendless Nice! Few points:

1. How does this play with the settings the user set on a language itself? Wouldn't your code you did get overloaded by the language settings anyway? I just tried putting this non-breaking in those fields and it solved the issue, so users would have to manually edit their date formats anyway, WDYT?

I am thinking of going the other way. What if we wrap date and time to two spans and set them to no-break? That would make it universal, regardless of where the date format comes from. 🤔

![Výstřižek](https://user-images.githubusercontent.com/6097524/125359930-5c088480-e36b-11eb-9a8b-f35687f8260d.JPG)

____________________

2. Is there a possibility that a date with non breaking hyphen would end up in some SQL query somewhere? Is there a danger in that?

__________________

3. Just tested other formats and it seems to break correctly. 😎
![hyphen](https://user-images.githubusercontent.com/6097524/125359965-69be0a00-e36b-11eb-91cb-9fb1ca819330.JPG)
![dot](https://user-images.githubusercontent.com/6097524/125359969-6aef3700-e36b-11eb-8f57-38a84a497112.JPG)

PierreRambaud reviewed (dismissed):
Why you don't use a constant? or a sanitize Twig helper to do it.
I'm afraid if people create a new pull request, they will use standard hyphens instead of non-breaking one.

Commit:
Wrap content of DateTimeColumns in <time> tags

Commit:
Use contextual date format in Log grid

Commit:
Transform hyphens into non-breaking hyphens when feeding a different format to DateTimeColumn

eternoendless commented:
I updated the PR and its description.

- The content of DateTimeColumns is now wrapped in `<time>` tags and formatted according to whatever's configured in Language (this was missing in the Logs grid, which was using a hardcoded format)
- Since these formats vary with language (as defined in https://github.com/PrestaShop/PrestaShop/blob/develop/app/Resources/all_languages.json) and can be user-edited, I added a little normalization to replace dashes with non breaking hyphens in DateTimeColumn

> What if we wrap date and time to two spans and set them to no-break?

I thought about this but it's incompatible with user-defined formats, where the date and the time are in the same format. I was unable to find a clean way to workaround that.

> Is there a possibility that a date with non breaking hyphen would end up in some SQL query somewhere? Is there a danger in that?

It's not impossible that someone somewhere tried to use the formats in Language to format SQL input, but honestly, if they did they were looking for trouble — these are meant for output and vary with language.

matks commented on src/Core/Grid/Definition/Factory/LogGridDefinitionFactory.php:
```diff
@@ -49,6 +50,19 @@ final class LogGridDefinitionFactory extends AbstractGridDefinitionFactory
+        string $contextDateFormat
+    ) {
+        parent::__construct($hookDispatcher);
+        $this->contextDateFormat = $contextDateFormat;
```
New constructor argument: BC break?

Progi1984 replied:
Negative. The class is final so can't be inherited.

eternoendless replied:
I will note it in the rare case someone would instantiate this class manually.

NeOMakinG commented on templates/bundles/PrestaShopBundle/Admin/Common/Grid/Columns/Content/date_time.html.twig:
```diff
@@ -23,10 +23,8 @@
-  {% set valueToDisplay = column.options.empty_data %}
+  {{  column.options.empty_data }}
+{% else %}
+  <time datetime="{{ record[column.id]|date(constant('PrestaShop\\PrestaShop\\Core\\Util\\DateTime\\DateTime::ISO_DATETIME_FORMAT')) }}">{{ record[column.id]|date(column.options.format) }}</time>
```
The `time` element is not supported at all on IE, this means that date fields will be totally empty on IE :/

I'm not in favor of supporting IE anymore, but this should be decided and communicated before using non-supported features, wdyt?

kpodemski replied:
```
<!--[if IE]>please, update it finally 😂 <![endif]-->
```

PierreRambaud replied:
IE is dead, if it works on edge, that should be enough as we target the BO and not the FO

NeOMakinG replied:
Even on FO tbh

khouloudbelguith commented:
Hi @eternoendless,

Could you please check the Conflicting files?

Thank you!

eternoendless commented:
Rebased

HanaRebaiQA commented:
Hello,

I verified this PR. The issue is fixed.
I tested the behavior in different pages which contain date with : 
different languages (English, Frensh, Arabe), different date format (`- , : ; _ /`), different browsers (Chrome, Firefox, Safari) in different devices (IOS, Android)

LGTM, QA :heavy_check_mark: 

See screen-record : 
https://www.awesomescreenshot.com/video/6035556?key=936182a6f1bc5a07de2bc5dab2b29460

Thanks,

Progi1984 commented:
Thanks @eternoendless & @HanaRebaiQA
